Allow loading a model SDF file in the Server class

JoanAguilar opened the pull request:
# 🎉 New feature

Closes #1711

## Summary
The constructor of the Server class has been updated to handle SDF files that contain a model without a world. The model is loaded into the default world.

## Test it
One should be able to create Server instances using model SDF files, for instance: [`test/worlds/models/sphere/model.sdf`](https://github.com/gazebosim/gz-sim/blob/gz-sim7/test/worlds/models/sphere/model.sdf).

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

mjcarroll commented:
Since `UNIT_Server_TEST` is already pretty close to the timeout when run in Github Actions (at least, maybe other places as well), I'm a little concerned about adding tests that also depend on rendering as well.

Do you think that you could break this out into UNIT_Server_Rendering_TEST to make it a separate executable and also contain the rendering dep?

JoanAguilar commented:
> Do you think that you could break this out into UNIT_Server_Rendering_TEST to make it a separate executable and also contain the rendering dep?

OK, sure. Note that (for now) there's only going to be one test in the new executable: `LoadSdfModelRelativeUri`.

JoanAguilar commented on src/Server.cc:
```diff
@@ -113,12 +113,31 @@ Server::Server(const ServerConfig &_config)
+        {
+          // If the specified file only contains a model, load the default
+          // world and add the model to it.
+          const sdf::Model model = *(sdfRoot.Model());
```
I'll revisit this.

mjcarroll commented:
> Note that (for now) there's only going to be one test in the new executable: LoadSdfModelRelativeUri.

Yes, I think that will be fine.

JoanAguilar commented on src/Server.cc:
```diff
@@ -113,12 +113,31 @@ Server::Server(const ServerConfig &_config)
+          // world and add the model to it.
+          const sdf::Model model = *(sdfRoot.Model());
+          errors = this->dataPtr->sdfRoot.LoadSdfString(DefaultWorld::World());
+          assert(this->dataPtr->sdfRoot.WorldCount() == 1);
```
Yeah, I was worndering about that. The other options that come to mind would be to catch this via if statement, or leave this uncaught. I have no strong preference, I'll go with whatever you-all think is best.

JoanAguilar commented:
Previous comments should be addressed now:

 - `LoadSdfModelRelativeUri` is now on its own file.

 - I have removed the `assert`s.

 - I have updated `FindScene` to use gtest macros to determine success (rather than returning a boolean).
 
 - Some other minor changes.

mjcarroll commented:
Looking better, I think that Windows and macOS are having failures in the test that you added.  The Ubuntu failure is known and can be ignored.

azeey commented:
Looks like the windows tests are failing on the stable CI as well https://build.osrfoundation.org/job/ign_gazebo-gz-sim7-win/17/

mjcarroll commented:
Oops, you are right, I misread that.

Also, I'm interested in why these compiler warnings have all appeared in the Windows build here, I thought that we just addressed that via the #1771

azeey commented:
> Also, I'm interested in why these compiler warnings have all appeared in the Windows build here, I thought that we just addressed that via the https://github.com/gazebosim/gz-sim/pull/1771

This branch needs to merge from `gz-sim7`. Unfortunately, when we enabled the ability to merge a PR without being up-to-date with the target branch, github removed the "Update branch" button.

azeey commented on src/Server_Rendering_TEST.cc:
```diff
@@ -0,0 +1,156 @@
+};
+
+/////////////////////////////////////////////////
+TEST_P(ServerFixture, LoadSdfModelRelativeUri)
```
Looks like we are not running tests that require rendering on macOS. Can you disable it for macOS for now?
```suggestion
TEST_P(ServerFixture, GZ_UTILS_TEST_DISABLED_ON_MAC(LoadSdfModelRelativeUri))
```

See 
https://github.com/gazebosim/gz-sim/blob/acd016a3ae18997443c8695213f918d44c68e638/test/integration/rgbd_camera.cc#L64

azeey commented on src/CMakeLists.txt:
```diff
@@ -197,6 +198,13 @@ gz_build_tests(TYPE UNIT
 )
 
+# Some server unit tests require rendering.
+if (TARGET UNIT_Server_Rendering_TEST)
```
We should also add this target to the `tests_needing_display` list

JoanAguilar commented on src/Server_TEST.cc:
```diff
@@ -519,6 +519,18 @@ TEST_P(ServerFixture, SdfStringServerConfig)
 }
 
+/////////////////////////////////////////////////
+TEST_P(ServerFixture, LoadSdfModel)
```
Fine with me. This test will be in `Server_Rendering_TEST.cc` and not require rendering, but I understand we are OK with that.

JoanAguilar commented:
I pushed another batch of changes:

 - I added `Server_Rendering_TEST.cc` to the list of `tests_needing_display`.

 - Per [comment](https://github.com/gazebosim/gz-sim/pull/1775#discussion_r1014268027), I disabled `LoadSdfModelRelativeUri` for macOS.

 - I moved `LoadSdfModel` to `Server_Rendering_TEST.cc`. I think there is some value in keeping it (for instance, `LoadSdfModel` will run on macOS, while `LoadSdfModelRelativeUri` won't); however, it does feel a bit misplaced.
 
 - I made tests on `Server_Rendering_TEST.cc` run only once.

CI looks a lot more green now, but that might be in part because `LoadSdfModelRelativeUri` does not seem to run in any of the builds. Looking at the logs, I keep seeing the lines

```
Looking for display capabilities
 ! valid display not found
 ! valid dri display not found (no glxinfo or pyopengl)
```

and I don't see references to `UNIT_Server_Rendering_TEST` (I do see references to `UNIT_Server_TEST`).

---

As a side note: after I moved `UNIT_Server_Rendering_TEST` to the list of `tests_needing_display` it doesn't run anymore on my local build. I wonder if the fact that I'm using Docker + rocker is interfering with the logic in charge of detecting a valid display (display capabilities seem to work fine, as I'm able to launch a GUI windows). And talking about GUI windows, there also seems to be a set of GUI tests that are not in the `tests_needing_display` list, but that run regardless (and spawn new windows while they run).

mjcarroll commented:
> As a side note: after I moved UNIT_Server_Rendering_TEST to the list of tests_needing_display it doesn't run anymore on my local build

It's likely that the detection is broken because of missing application.  I believe that you need to have `env DISPLAY` set, and then you need `xwininfo` from `x11-utils` and `glxinfo` from `mesa-utils`.

Currently, these things will come back as "STATUS" messages, but maybe we can elevate them to warning to make it more obvious.
